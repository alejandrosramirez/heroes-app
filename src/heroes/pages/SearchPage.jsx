import { useLocation, useNavigate } from "react-router-dom";
import queryString from "query-string";
import { useForm } from "../../hooks";
import { HeroCard } from "../components";
import { getHeroesByName } from "../helpers";

export const SearchPage = () => {
	const navigate = useNavigate();
	const location = useLocation();

	const { q = "" } = queryString.parse(location.search);

	const heroes = getHeroesByName(q);

	const showSearch = q.length === 0;
	const showError = q.length > 0 && heroes.length === 0;

	const { onChange, search } = useForm({
		search: q,
	});

	const handleSearchSubmit = (e) => {
		e.preventDefault();

		navigate(`?q=${search.toLowerCase().trim()}`);
	};

	return (
		<>
			<h1>Search Page</h1>

			<hr />

			<div className="row">
				<div className="col-5">
					<h4>Searching</h4>

					<hr />

					<form onSubmit={handleSearchSubmit} aria-label="form">
						<input
							className="form-control"
							type="text"
							placeholder="Search a hero..."
							name="search"
							autoComplete="off"
							value={search}
							onChange={onChange}
						/>

						<button className="btn btn-outline-primary mt-1">
							Search
						</button>
					</form>
				</div>

				<div className="col-7">
					<h4>Results</h4>

					<hr />

					{/* {
						q === ""
							?
							<div className="alert alert-primary">
								Search a hero...
							</div>
							: (heroes.length === 0) && (
								<div className="alert alert-danger">
									There's no hero with <b>{q}</b>
								</div>
							)
					} */}

					<div
						className="alert alert-primary animate__animated animate__fadeInLeft"
						style={{ display: showSearch ? "" : "none" }}
					>
						Search a hero...
					</div>

					<div
						className="alert alert-danger animate__animated animate__fadeInLeft"
						style={{ display: showError ? "" : "none" }}
						aria-label="not-found-hero-alert"
					>
						There's no hero with <b>{q}</b>
					</div>

					{heroes.map((hero) => (
						<HeroCard key={hero.id} {...hero} />
					))}
				</div>
			</div>
		</>
	);
};
