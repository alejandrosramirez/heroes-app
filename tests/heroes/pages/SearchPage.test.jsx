import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { SearchPage } from "../../../src/heroes/pages";

const mockUseNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
	...jest.requireActual("react-router-dom"),
	useNavigate: () => mockUseNavigate,
}));

describe("Pruebas en <SearchPage />", () => {
	beforeEach(() => jest.clearAllMocks());

	test("Debe de mostrarse correctamte con valores por defecto", () => {
		const { container } = render(
			<MemoryRouter>
				<SearchPage />
			</MemoryRouter>
		);
		expect(container).toMatchSnapshot();
	});

	test("Debe de mostrar a Batman y el input con el valor del query string", () => {
		render(
			<MemoryRouter initialEntries={["/search?q=batman"]}>
				<SearchPage />
			</MemoryRouter>
		);
		const input = screen.getByRole("textbox");
		expect(input.value).toBe("batman");
		const img = screen.getByRole("img");
		expect(img.src).toContain("/assets/heroes/dc-batman.jpg");
		const searchHeroAlert = screen.getByLabelText("not-found-hero-alert");
		expect(searchHeroAlert.style.display).toBe("none");
	});

	test("Debe de mostrar un error si no se encuentra un hÃ©roe (batman123).", () => {
		render(
			<MemoryRouter initialEntries={["/search?q=batman123"]}>
				<SearchPage />
			</MemoryRouter>
		);
		const notFoundHeroAlert = screen.getByLabelText("not-found-hero-alert");
		expect(notFoundHeroAlert.style.display).toBe("");
	});

	test("Debe llamar el navigate a la pantalla nueva", () => {
		const value = "superman";
		render(
			<MemoryRouter initialEntries={["/search"]}>
				<SearchPage />
			</MemoryRouter>
		);
		const input = screen.getByRole("textbox");
		fireEvent.change(input, { target: { name: "search", value } });
		const form = screen.getByRole("form");
		fireEvent.submit(form);
		expect(mockUseNavigate).toHaveBeenCalledWith(`?q=${value}`);
	});
});
