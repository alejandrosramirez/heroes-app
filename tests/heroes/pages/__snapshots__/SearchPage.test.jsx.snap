// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchPage /> Debe de mostrarse correctamte con valores por defecto 1`] = `
<div>
  <h1>
    Search Page
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="search"
          placeholder="Search a hero..."
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        class="alert alert-primary animate__animated animate__fadeInLeft"
      >
        Search a hero...
      </div>
      <div
        aria-label="not-found-hero-alert"
        class="alert alert-danger animate__animated animate__fadeInLeft"
        style="display: none;"
      >
        There's no hero with 
        <b />
      </div>
    </div>
  </div>
</div>
`;
